core/templates/pages/exploration-editor-page/services/router.service.spec.ts

```diff
@@ -420,7 +409,22 @@ describe('Router Service', function() {
   it('should save pending changes', function() {
     var broadcastSpy = spyOn($rootScope, '$broadcast').and.callThrough();
     RouterService.savePendingChanges();
-    $timeout.flush();
     expect(broadcastSpy).toHaveBeenCalledWith('externalSave');
   });
+
+  it('should save pending changes even when AngularJS throws an error',
+    function() {
+      spyOn($rootScope, '$broadcast').and.throwError(
+        'Cannot read property $$nextSibling of null');
+      var applySpy = spyOn($rootScope, '$apply').and.callThrough();
+      // $rootScope.$destroy is being mocked because it is called
+      // in AngularJS flow when savePendingChanges() is executed.
+      // This method calls $broadcast, that will throw an error on
+      // Angular because of spyOn().
+      // Ref: Check third_party/static/angularjs-1.7.9/angular.js:19338
+      spyOn($rootScope, '$destroy').and.callFake(function() {});
+
+      RouterService.savePendingChanges();
+      expect(applySpy).toHaveBeenCalled();
```

Comment:
I think this seems to be the wrong thing to check if what you want to verify is that pending changes have been saved. Could you find a different way to verify that pending changes have been saved, that is actually a reliable indicator of that? E.g. check that the version has changed, or that some data has changed, or something like that? The issue is that, in general, $rootScope.$apply could be called for a whole bunch of different things and there is no guarantee that the call to it corresponds to the saving of pending changes.

Reply:
I've changed it, I'm spying $broadcast now instead of $apply.

Reply:
I think it's not possible to check if a version has changed or something like that because the only thing that is called in this function is $broadcast.

Reply:
Hm. I looked into the implementation of savePendingChanges and the main thing it does it to broadcast an externalSave event. So, the test should be checking whether this externalSave event is actually broadcast, or is actually received by other controllers. We should be seeing 'externalSave' somewhere in the test. Would that be possible?

Reply:
Done.